net: tcp: Implement TCP new Reno collision avoidance

ssharks opened the pull request:
To avoid a TCP connection from collapsing a link, implement a collision avoidance algorithm. Initially TCP new Reno is implemented for its simplicity.

ssharks commented:
Attempts to implement: https://github.com/zephyrproject-rtos/zephyr/issues/46352

ssharks committed:
net: tcp: Implement TCP new Reno collision avoidance

To avoid a TCP connection from collapsing a link, implement a collision
avoidance algorithm. Initially TCP new Reno is implemented for its
simplicity.

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

ssharks commented:
@rlubos 
During the style check I get:
0dc74b0e2f13add650033e25396c7adc37305c30:45: ERROR:API_DEFINE: do not specify a non-Zephyr API for libc
#45: FILE: subsys/net/ip/tcp.c:66:+#define TCP_CONGESTION_INITIAL_WIN 1
Do you know what I should do in order to prevent this style violation?

rlubos commented:
@ssharks Looks like a checkpatch configuration bug I've seen in other PR as well, a fix has been submitted: https://github.com/zephyrproject-rtos/zephyr/pull/60551 - can you rebase and see if it helps?

ssharks commented:
> To avoid a TCP connection from collapsing a link, implement a collision avoidance algorithm. Initially TCP new Reno is implemented for its simplicity.

Other congestion avoidance algorithm like the Linux default CUBIC will performance better on long fat pipes. As the Zephyr TCP stack support only 64K transmit window, the difference is likely neglectable even for long latencies.

ssharks commented:
Please have a look at the form factor. Currently I've put it all in tcp.c if we would put it in a seperate file it would be possible to switch between different algorithms compile time. The specific CA algorithm will just fill in the functions.

I see I need some small changes to properly implement the rfc3782

ssharks committed:
net: tcp: Log the steps in the collision avoidance

To allow insighed into the correct functioning of the collision avoidance,
log the internal values and function calls.

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

fgrandel reviewed (commented):
Looks like a very useful addition - thx! Unfortunately don't know enough for a full review. Just one minor suggestion.

rlubos commented on subsys/net/ip/tcp.c:
```diff
@@ -392,6 +396,47 @@ static void tcp_derive_rto(struct tcp *conn)
+
+static void tcp_new_reno_fast_retransmit(struct tcp *conn)
+{
+	conn->ca.congestion_win = MAX(conn_mss(conn), conn->ca.congestion_win / 2);
```
Could you reference the RFC/chapter where the formulas come from? It should simplify future maintenance.

ssharks replied:
Done

ssharks commented:
During testing I bumped into a corner case, which gave me random failing unit tests. A PR for this issue is submitted at:
https://github.com/zephyrproject-rtos/zephyr/pull/60880

ssharks commented on subsys/net/ip/tcp.c:
```diff
@@ -392,6 +396,47 @@ static void tcp_derive_rto(struct tcp *conn)
+
+	if (conn->ca.state == TCP_NEW_RENO_RAMPUP) {
+		new_win += conn_mss(conn);
+		conn->ca.congestion_win += conn_mss(conn);
```
Code has fully been reworked here

ssharks commented on subsys/net/ip/tcp.c:
```diff
@@ -1306,6 +1375,15 @@ static void tcp_resend_data(struct k_work *work)
 		goto out;
 	}
 
+#ifdef CONFIG_NET_TCP_CONGESTION_AVOIDANCE
```
Fixed

ssharks commented on subsys/net/ip/tcp.c:
```diff
@@ -1162,6 +1222,13 @@ static int tcp_unsent_len(struct tcp *conn)
 		unsent_len = 0;
 	} else {
 		unsent_len = MIN(unsent_len, conn->send_win - conn->unacked_len);
+#ifdef CONFIG_NET_TCP_CONGESTION_AVOIDANCE
```
See comment below

ssharks commented on subsys/net/ip/tcp.c:
```diff
@@ -1140,6 +1197,9 @@ static int tcp_pkt_peek(struct net_pkt *to, struct net_pkt *from, size_t pos,
 static bool tcp_window_full(struct tcp *conn)
 {
 	bool window_full = (conn->send_data_total >= conn->send_win);
+#ifdef CONFIG_NET_TCP_CONGESTION_AVOIDANCE
```
Fixed

fgrandel reviewed (commented):
LGTM now wrt my superficial remark above, still no full review, though, as I don't understand enough of what's going on there conceptually.

rlubos reviewed (commented):
Looks good, just some nits. Did you have a chance to run any performance test (like `zperf`) to see what is the impact on the actual throughput? I guess it should be negligible but it would be good to see if that's really the case.

ssharks commented on subsys/net/ip/tcp.c:
```diff
@@ -392,6 +396,127 @@ static void tcp_derive_rto(struct tcp *conn)
+
+static void tcp_ca_pkts_acked(struct tcp *conn, uint32_t acked_len)
+{
+}
```
Thanks done

ssharks commented on subsys/net/ip/Kconfig:
```diff
@@ -456,6 +456,14 @@ config NET_TCP_FAST_RETRANSMIT
 	  the retransmit timer to elapse.
 
+config NET_TCP_CONGESTION_AVOIDANCE
+	bool "Implement a collision avoidance algorithm in TCP"
```
Oops, fix attempt 2

ssharks commented:
> Looks good, just some nits. Did you have a chance to run any performance test (like `zperf`) to see what is the impact on the actual throughput? I guess it should be negligible but it would be good to see if that's really the case.

With respect to throughput, the only thing I can see is that the unit test does not get significantly slower. I do not have HW which I can run a zperf on saidly, so cannot test that.

ssharks commented:
Performed a rebase to trigger a rebuild of the CI/CD

ssharks commented:
@jukkar: 
During testing, I bumped into an issue regarding line: https://github.com/zephyrproject-rtos/zephyr/blame/203b474c4b2f51927c13c83b55a85df1ef6bc4a6/subsys/net/ip/tcp2.c#L2186C9-L2186C34

When the transmit window is full, which happens much more often with congestion avoidance, the resend timer is being triggered by new transmissions. This causes a lot of spurious retransmissions.
It looks like you have created this line, do you still remember the purpose of it?

ssharks commented:
> @jukkar: During testing, I bumped into an issue regarding line:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/488fd89e2bd64c60354378fdec33b7e65c713d38/subsys/net/ip/tcp.c#L2795-L2797
> 
> When the transmit window is full, which happens much more often with congestion avoidance, the resend timer is being triggered by new transmissions. This causes a lot of spurious retransmissions. It looks like you have created this line, do you still remember the purpose of it?

A PR to remove this is already available: https://github.com/zephyrproject-rtos/zephyr/pull/60972

rlubos commented:
@ssharks FYI, I've checked with `zperf` (1s/10s throughput test) and I see no noticable changes in the TCP throughput.

jukkar reviewed (approved):
LGTM
